
<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.19: https://docutils.sourceforge.io/" />

    <title>Get variants associated with a trait &#8212; genopyc 2.3.0 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/alabaster.css" />
    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/sphinx_highlight.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="Installation" href="installation.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="get-variants-associated-with-a-trait">
<h1>Get variants associated with a trait<a class="headerlink" href="#get-variants-associated-with-a-trait" title="Permalink to this heading">¶</a></h1>
<p><strong>Genopyc allow the user of querying GWAS catalog to query for SNPs associated to a complex trait. The trait has to be expressed as experimental factor ontology ID.</strong></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">get_associations</span><span class="p">(</span><span class="n">efotrait</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">studyid</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
<p>Retrieve SNPs associated with an EFO trait.</p>
<p>Parameters:</p>
<ul class="simple">
<li><p>efotrait (str): EFO id of the trait for which variants are to be retrieved (ex. EFO_0004994)</p></li>
<li><p>verbose (bool, optional): If True, the function returns the steps of variant retrieval</p></li>
<li><p>studyid (bool, optional): If True, the Id of the GWAS where the association was found is retrieved</p></li>
</ul>
<p>Returns:</p>
<ul class="simple">
<li><p>Pandas DataFrame: Variants associated with the trait</p></li>
</ul>
<p><strong>Once the SNPs are retrieved they can be investigated with the function:</strong></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">get_variants_info</span><span class="p">(</span><span class="n">idlist</span><span class="p">,</span> <span class="n">chunked</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">chunksize</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
</pre></div>
</div>
<p>Retrieve information about variants in the Homo sapiens species from the Ensembl variation database.</p>
<p>Parameters:</p>
<ul class="simple">
<li><p>idlist (list or str): A list of variant IDs or a single variant ID as a string.</p></li>
<li><p>chunked (bool, optional): Whether to chunk the requests into smaller portions if the number of IDs exceeds 200. Default is False.</p></li>
<li><p>chunksize (int, optional): The size of each chunk when chunking requests. Default is 200.</p></li>
</ul>
<p>Returns:</p>
<ul class="simple">
<li><p>dict: A dictionary containing information about the variants, where keys are variant IDs and values are variant information.</p></li>
</ul>
<p>Note:
This function requires internet connectivity to access the Ensembl REST API.
Ensembl accepts a maximum of 200 IDs per request. If the number of IDs exceeds 200 and chunking is not disabled,
the requests will be split into smaller chunks to retrieve variant information.</p>
<p><strong>Moreover genopyc allows to retrieve genes or variants associated to a specific trait throug DisGeNET one of the largest repositories of genes and variants associated to human diseases.</strong></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">get_gdas</span><span class="p">(</span><span class="n">query_list</span><span class="p">,</span> <span class="n">username</span><span class="p">,</span> <span class="n">password</span><span class="p">,</span> <span class="n">mode</span><span class="p">)</span>
</pre></div>
</div>
<p>Retrieve gene or variant-disease associations from the DisGeNET database.</p>
<p>Parameters:</p>
<ul class="simple">
<li><p>query_list (list): A list of gene symbols or variant IDs for which associations are requested.</p></li>
<li><p>username (str): Username for accessing the DisGeNET API.</p></li>
<li><p>password (str): Password for accessing the DisGeNET API.</p></li>
<li><p>mode (str): Mode of query, either ‘genes’ or ‘variants’.</p></li>
</ul>
<p>Returns:</p>
<ul class="simple">
<li><p>pandas DataFrame: A DataFrame containing gene or variant-disease associations.</p></li>
</ul>
<p>Note:
This function queries the DisGeNET <sup>1</sup> API to retrieve gene or variant-disease associations.
It requires authentication using a username and password provided by the user.
The mode parameter specifies whether the query is for genes or variants.</p>
</section>
<section id="investigate-genomic-locations">
<h1>Investigate genomic locations<a class="headerlink" href="#investigate-genomic-locations" title="Permalink to this heading">¶</a></h1>
<p><strong>Genopyc includes many functions to investigate genomic locations, specifically, given a list of variants the user can retrieve the genomic locations:</strong></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">get_variants_position</span><span class="p">(</span><span class="n">idlist</span><span class="p">,</span> <span class="n">chunked</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">chunksize</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
</pre></div>
</div>
<p>Retrieve the positions (chromosome and position) of variant IDs from the Ensembl database for Homo sapiens.</p>
<p>Parameters:</p>
<ul class="simple">
<li><p>idlist (list or str): A list of variant IDs or a single variant ID as a string.</p></li>
<li><p>chunked (bool, optional): Whether to chunk the requests into smaller portions. Default is False.</p></li>
<li><p>chunksize (int, optional): The size of each chunk when chunking requests. Default is 200.</p></li>
</ul>
<p>Returns:</p>
<ul class="simple">
<li><p>list of tuples: A list of tuples containing the variant ID, chromosome, and position.</p></li>
</ul>
<p>Note:
This function requires internet connectivity to access the Ensembl REST API.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">get_genes_position</span><span class="p">(</span><span class="n">idlist</span><span class="p">,</span> <span class="n">chunked</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">chunksize</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>

</pre></div>
</div>
<p>Retrieve the coordinates of genes in the GRCh38 assembly for a list of Ensembl IDs.</p>
<p>Parameters:</p>
<ul class="simple">
<li><p>idlist (list or str): A list of Ensembl gene IDs or a single Ensembl gene ID as a string.</p></li>
<li><p>chunked (bool, optional): Whether to chunk the requests into smaller portions if the number of IDs exceeds 200. Default is False.</p></li>
<li><p>chunksize (int, optional): The size of each chunk when chunking requests. Default is 200.</p></li>
</ul>
<p>Returns:</p>
<ul class="simple">
<li><p>list of tuples: A list of tuples containing the Ensembl gene ID, chromosome number, start position, and end position.</p></li>
</ul>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">get_genes</span><span class="p">(</span><span class="n">cr</span><span class="p">,</span> <span class="n">location</span><span class="p">,</span> <span class="n">window_size</span><span class="o">=</span><span class="mi">10000</span><span class="p">,</span> <span class="n">pop</span><span class="o">=</span><span class="s1">&#39;EUR&#39;</span><span class="p">,</span> <span class="n">features</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;gene&#39;</span><span class="p">],</span> <span class="n">mode</span><span class="o">=</span><span class="s1">&#39;all&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>Retrieve genes in a window centered around a genomic position and compute the distance between the position and all genes.</p>
<p>Parameters:</p>
<ul class="simple">
<li><p>cr (str): Chromosome identifier.</p></li>
<li><p>location (int): Genomic position around which the window is centered.</p></li>
<li><p>window_size (int, optional): Size of the window in base pairs. Default is 10,000.</p></li>
<li><p>pop (str, optional): Population for which to retrieve gene data. Default is ‘EUR’ (European).</p></li>
<li><p>features (list, optional): List of features to include in the retrieval. Default is [‘gene’].</p></li>
<li><p>mode (str, optional): Retrieval mode. Options: ‘all’ (returns all genes and their distances),
‘complete_data’ (returns complete response from the API), ‘closest_forward’ (returns the closest gene
located forward of the position), ‘closest_backward’ (returns the closest gene located backward of
the position), ‘closest_overall’ (returns the closest gene regardless of direction). Default is ‘all’.</p></li>
</ul>
<p>Returns:</p>
<ul class="simple">
<li><p>dict or str: Depending on the mode parameter:</p></li>
<li><p>If mode is ‘all’, returns a dictionary where keys are gene names and values are their distances from the position.</p></li>
<li><p>If mode is ‘complete_data’, returns the complete response from the API.</p></li>
<li><p>If mode is ‘closest_forward’ or ‘closest_backward’, returns the name of the closest gene in the specified direction.</p></li>
<li><p>If mode is ‘closest_overall’, returns the name of the closest gene regardless of direction.</p></li>
<li><p>If no genes are found, returns a message indicating the absence of genes.</p></li>
</ul>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">get_ov_region</span><span class="p">(</span><span class="n">snp</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="nb">chr</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">location</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">window</span><span class="o">=</span><span class="mi">500</span><span class="p">,</span> <span class="n">features</span><span class="o">=</span><span class="nb">list</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s1">&#39;region&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>Retrieve overlap regions for a specified SNP or genomic location.</p>
<p>Parameters:</p>
<ul class="simple">
<li><p>snp (str, optional): The variant ID (SNP) for which overlap regions are requested. Default is None.</p></li>
<li><p>chr (str or int, optional): Chromosome name or identifier. Required if mode is ‘SNP’.</p></li>
<li><p>location (int, optional): Genomic location (start position) for which overlap regions are requested. Required if mode is ‘region’.</p></li>
<li><p>window (int, optional): Window size to extend the genomic region around the SNP or location. Default is 500.</p></li>
<li><p>features (list, optional): A list of genomic feature types to include in the overlap regions. Default is an empty list.</p></li>
<li><p>mode (str, optional): Mode of operation, either ‘SNP’ or ‘region’. Default is ‘region’.</p></li>
</ul>
<p>Returns:</p>
<ul class="simple">
<li><p>list of pandas DataFrames: A list of pandas DataFrames containing overlap regions for each specified genomic feature type.</p></li>
</ul>
<p>Note:
This function queries the Ensembl REST API to retrieve overlap regions for the specified SNP or genomic location.
It returns a list of pandas DataFrames, where each DataFrame contains overlap regions for a specific genomic feature type.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">get_sequence</span><span class="p">(</span><span class="n">chromosome</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">stop</span><span class="p">)</span>
</pre></div>
</div>
<p>Retrieve the genomic sequence for a given genomic region.</p>
<p>Parameters:</p>
<ul class="simple">
<li><p>chromosome (str or int): Chromosome name or identifier.</p></li>
<li><p>start (int): Start position of the genomic region.</p></li>
<li><p>stop (int): End position of the genomic region.</p></li>
</ul>
<p>Returns:</p>
<ul class="simple">
<li><p>str: The genomic sequence for the specified genomic region.</p></li>
</ul>
<p>Note:
This function queries the Ensembl REST API to retrieve the genomic sequence for the specified genomic region.
It returns the genomic sequence as a string.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">closest_genes</span><span class="p">(</span><span class="n">position_id</span><span class="p">,</span> <span class="n">chromosome</span><span class="p">,</span> <span class="n">position</span><span class="p">,</span> <span class="n">window_size</span><span class="p">,</span> <span class="n">type_of_gene</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s1">&#39;start&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>Find the closest upstream and downstream genes to a given genomic position.</p>
<p>Parameters:</p>
<ul class="simple">
<li><p>position_id (str): Identifier for the genomic position.</p></li>
<li><p>chromosome (str or int): Chromosome name or identifier.</p></li>
<li><p>position (int): Genomic position for which closest genes are to be found.</p></li>
<li><p>window_size (int): Window size to extend the genomic region around the position.</p></li>
<li><p>type_of_gene (str, optional): Biotype of genes to consider. Default is False (include all genes).</p></li>
<li><p>mode (str, optional): Mode of calculating distances, either ‘start’ or ‘absolute’. Default is ‘start’.
start:  In this case the distance will be calculated from the beginning of both upstream and downstream genes
absolute: Upstream distance is calcuated as distance (base pairs) to the end and downstream as distance to the start of the gene</p></li>
</ul>
<p>Returns:</p>
<ul class="simple">
<li><p>tuple: A tuple containing the position ID, the closest upstream gene ID, and the closest downstream gene ID.</p></li>
</ul>
<p>Note:
It calculates distances from the specified position to the start (or end) of genes and finds the closest upstream and downstream genes.</p>
</section>
<section id="linkage-disequilibrium">
<h1>Linkage Disequilibrium<a class="headerlink" href="#linkage-disequilibrium" title="Permalink to this heading">¶</a></h1>
<p><strong>Linkage disequilibrium (LD) is an important aspect when it comes to investigate the effect of variants on downstream biological pathways. In fact many times a SNPs significantly correlated with a phenotype through GWAS is associated with many others due to this phenomena. Thus, determining which is the causative SNP is not a trivial task. With genopyc information related to LD in different ways:</strong></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">get_variants_in_LD</span><span class="p">(</span><span class="n">variant</span><span class="p">,</span> <span class="n">r2</span><span class="p">,</span> <span class="n">pop</span><span class="o">=</span><span class="s1">&#39;EUR&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>Retrieve variants in linkage disequilibrium (LD) with the specified variant.</p>
<p>Parameters:</p>
<ul class="simple">
<li><p>variant (str): The variant ID for which LD variants are requested.</p></li>
<li><p>r2 (float): The minimum threshold for the linkage disequilibrium coefficient (r2).</p></li>
<li><p>pop (str, optional): The population for which LD information is requested. Default is ‘EUR’ (European population).</p></li>
</ul>
<p>Returns:</p>
<ul class="simple">
<li><p>list: A list of variant IDs that are in linkage disequilibrium with the specified variant, based on the given r2 threshold.</p></li>
</ul>
<p>Note:
This function queries the Ensembl REST API to retrieve LD information for the specified variant.
It returns a list of variant IDs that are in LD with the specified variant, with an r2 coefficient greater than or equal to the specified threshold.
If no variants are found or an error occurs during retrieval, an empty list is returned.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">get_ld_matrix</span><span class="p">(</span><span class="n">list_of_snps</span><span class="p">,</span> <span class="n">token</span><span class="p">,</span> <span class="n">pop</span><span class="o">=</span><span class="s1">&#39;EUR&#39;</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="s1">&#39;r2&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>Get the linkage disequilibrium (LD) matrix for a list of SNPs.</p>
<p>Parameters:</p>
<ul class="simple">
<li><p>list_of_snps (list): A list of SNP IDs.</p></li>
<li><p>token (str): The authentication token for accessing the LDlink API.</p></li>
<li><p>pop (str, optional): The population for which LD information is requested. Default is ‘EUR’ (European population).</p></li>
<li><p>metric (str, optional): The LD metric to be used. Default is ‘r2’.</p></li>
</ul>
<p>Returns:</p>
<ul class="simple">
<li><p>pandas DataFrame: A DataFrame representing the LD matrix. Rows and columns are SNP IDs, and cell values represent the LD metric between corresponding SNPs.</p></li>
</ul>
<p>Note:
This function queries the LDlink API to retrieve the LD matrix for the specified list of SNPs, for this reason it needs the user to be registered on LDlink in order to obtain a token.
It returns a pandas DataFrame containing the LD matrix, where rows and columns are SNP IDs, and cell values represent the LD metric.
If any LD metric values are missing (NA), they are replaced with None, and the DataFrame is converted to numeric data types.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">pairwise_ld</span><span class="p">(</span><span class="n">ch</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">pop</span><span class="o">=</span><span class="s1">&#39;EUR&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>Retrieve pairwise linkage disequilibrium (LD) information for variants within a genomic region.</p>
<p>Parameters:</p>
<ul class="simple">
<li><p>ch (str): Chromosome name or identifier.</p></li>
<li><p>start (int): Start position of the genomic region.</p></li>
<li><p>end (int): End position of the genomic region.</p></li>
<li><p>pop (str, optional): The population for which LD information is requested. Default is ‘EUR’ (European population).</p></li>
</ul>
<p>Returns:</p>
<ul class="simple">
<li><p>pandas DataFrame: A DataFrame containing pairwise LD information between variants within the specified genomic region.
Columns include ‘v1’ and ‘v2’ for variant IDs and ‘r2’ for the LD coefficient.</p></li>
</ul>
<p>Note:
This function queries the Ensembl REST API to retrieve pairwise LD information for variants within the specified genomic region.
It returns a DataFrame containing pairwise LD information, where ‘v1’ and ‘v2’ represent variant IDs and ‘r2’ represents the LD coefficient.</p>
</section>
<section id="eqtls">
<h1>eQTLS<a class="headerlink" href="#eqtls" title="Permalink to this heading">¶</a></h1>
<p><strong>An important aspect of the possible imapct of the variants on the phenotype, expecially in relation to complex traits, is their effect on the expression of genes. This data can be gathered through genopyc with the functions:</strong></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">get_eqtl_df</span><span class="p">(</span><span class="n">rsid</span><span class="p">,</span> <span class="n">p_value</span><span class="o">=</span><span class="mf">0.005</span><span class="p">,</span> <span class="n">increase_index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
<p>Retrieve eQTL (expression quantitative trait loci) associations as a DataFrame for a given variant.</p>
<p>Parameters:</p>
<ul class="simple">
<li><p>rsid (str): The variant ID (e.g., rsID) for which eQTL associations are requested.</p></li>
<li><p>p_value (float, optional): Maximum p-value threshold for filtering eQTL associations. Default is 0.005.</p></li>
<li><p>increase_index (bool, optional): Whether to increase the DataFrame index by 1. Default is False.</p></li>
</ul>
<p>Returns:</p>
<ul class="simple">
<li><p>pandas DataFrame or None: A DataFrame containing eQTL associations for the specified variant if successful,
otherwise returns None.</p></li>
</ul>
<p>Note:
This function queries the EBI eQTL Catalog API to retrieve eQTL associations for the specified variant.
It returns a pandas DataFrame containing eQTL association information, including variant ID, p-value, beta value,
target gene ID, tissue, study ID, and tissue name. It optionally filters associations based on the provided p-value threshold.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span> <span class="n">get_eqtl_variant</span><span class="p">(</span><span class="n">rsid</span><span class="p">)</span>
</pre></div>
</div>
<p>Retrieve eQTL (expression quantitative trait loci) associations for a given variant.</p>
<p>Parameters:</p>
<ul class="simple">
<li><p>rsid (str): The variant ID (e.g., rsID) for which eQTL associations are requested.</p></li>
</ul>
<p>Returns:</p>
<ul class="simple">
<li><p>dict: A dictionary containing eQTL associations for the specified variant.</p></li>
</ul>
<p>Note:
This function queries the EBI eQTL Catalog API to retrieve eQTL associations for the specified variant.
It returns a dictionary containing eQTL association information, including target genes and other relevant data, for the specified variant.</p>
<p><strong>A functionality to express associations between variants and genes as network in a specific tissue is available in genopyc with the function</strong></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">make_eqtl_network</span><span class="p">(</span><span class="n">list_of_eqtls_df</span><span class="p">,</span> <span class="n">tissue</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">gene</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">variant</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">gene_symbol</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">tissue_name</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
</pre></div>
</div>
<p>Generate a network based on eQTL associations.</p>
<p>Parameters:</p>
<ul class="simple">
<li><p>list_of_eqtls_df (list): List of DataFrames containing eQTL information.</p></li>
<li><p>tissue (str, optional): Filter the network by tissue. Default is None.</p></li>
<li><p>gene (str, optional): Filter the network by gene. Default is None.</p></li>
<li><p>variant (list, optional): List of variants to include in the network. Default is None.</p></li>
<li><p>gene_symbol (bool, optional): Whether to use gene symbols instead of Ensembl IDs in the network. Default is False.</p></li>
<li><p>tissue_name (bool, optional): Whether to use tissue names instead of IDs in the network. Default is False.</p></li>
</ul>
<p>Returns:</p>
<ul class="simple">
<li><p>networkx.Graph: A networkx graph representing the eQTL network.</p></li>
</ul>
<p>Note:</p>
<ul class="simple">
<li><p>The function generates a (networkx) network based on eQTL associations from a list of DataFrames.</p></li>
<li><p>The network can be filtered by tissue, gene, or variant.</p></li>
<li><p>If ‘gene_symbol’ is True, gene symbols will be used instead of Ensembl IDs in the network.</p></li>
<li><p>If ‘tissue_name’ is True, tissue names will be used instead of IDs in the network.</p></li>
</ul>
<p><strong>In this network there are 3 types of nodes (genes, variants, and tissue) an edge exists if the variant is affecting the transcription of that specific gene in the tissue. Moreover, the edge carries the information of the strength of the association between the variant and the transcription increase or decrease of the gene. The network can be plotted as shown in the <a class="reference external" href="https://github.com/freh-g/genopyc/blob/main/tutorials/Genopyc_tutorial_notebook.ipynb">tutorial notebook</a>.</strong></p>
</section>
<section id="from-variants-to-genes">
<h1>From Variants to Genes<a class="headerlink" href="#from-variants-to-genes" title="Permalink to this heading">¶</a></h1>
<p><strong>Genopyc integrates many tools for investigating how the variants could act on genes. One already mentioned way for example is the gathering of eQTL data. To complement the analysis genopyc integrates also Ensembl variant effect predictor (VEP) <sup>2</sup></strong>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">VEP</span><span class="p">(</span><span class="n">idlist</span><span class="p">,</span><span class="n">input_type</span> <span class="o">=</span> <span class="s1">&#39;rsid&#39;</span><span class="p">,</span> <span class="n">chunked</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span><span class="n">chunksize</span><span class="o">=</span><span class="mi">200</span><span class="p">,</span><span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span><span class="n">all_data</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span><span class="n">plot</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span><span class="n">save_plot</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>Perform Variant Effect Prediction (VEP) using the Ensembl Variant Effect Predictor for a list of variant IDs.</p>
<p>Parameters:</p>
<ul class="simple">
<li><p>idlist (list or str): A list of variant IDs or a single variant ID as a string.</p></li>
<li><p>input_type (str, optional): Type of input IDs. Valid options are ‘rsid’ (default) or ‘hgvs’.</p></li>
<li><p>chunked (bool, optional): Whether to chunk the requests into smaller portions if the number of IDs exceeds 200. Default is False.</p></li>
<li><p>chunksize (int, optional): The size of each chunk when chunking requests. Default is 200.</p></li>
<li><p>verbose (bool, optional): Whether to print verbose information during processing. Default is False.</p></li>
<li><p>all_data (bool, optional): Whether to return all data retrieved from VEP. Default is False.</p></li>
<li><p>plot (bool, optional): Whether to generate a pie chart visualizing the consequences of the variants. Default is False.</p></li>
<li><p>save_plot (str, optional): Filepath to save the generated plot. If not provided, the plot will be displayed instead.</p></li>
</ul>
<p>Returns:</p>
<ul class="simple">
<li><p>list of DataFrames or dict: A list of DataFrames containing the VEP results (if <code class="docutils literal notranslate"><span class="pre">all_data</span></code> is False), or a dictionary containing all retrieved data (if <code class="docutils literal notranslate"><span class="pre">all_data</span></code> is True). if <code class="docutils literal notranslate"><span class="pre">plot</span></code> is True a pie chart with the effects of the variants is returned.</p></li>
</ul>
<p>Note:
This function requires internet connectivity to access the Ensembl REST API.</p>
<p><strong>Locus to gene (L2G) pipeline <sup>3</sup> from Open Target Genetics:</strong></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">OT_L2G</span><span class="p">(</span><span class="n">list_of_variants</span><span class="p">,</span> <span class="n">score</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">output</span><span class="o">=</span><span class="s1">&#39;genes&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>Retrieve genes associated with variants from the Open Targets Genetics API.</p>
<p>Parameters:</p>
<ul class="simple">
<li><p>list_of_variants (list): List of variant identifiers.</p></li>
<li><p>score (float, optional): Threshold score for gene association. Genes with scores higher than this threshold will be included. Default is 0.1.</p></li>
<li><p>output (str, optional): Output format. ‘genes’ returns a list of genes, ‘all’ returns a DataFrame with variant, gene, and score information. Default is ‘genes’.</p></li>
</ul>
<p>Returns:</p>
<ul class="simple">
<li><p>list or DataFrame: Depending on the output parameter, either a list of genes or a DataFrame with variant, gene, and score information.</p></li>
</ul>
<p>Note:</p>
<ul class="simple">
<li><p>The function retrieves genes associated with variants from the Open Targets Genetics API.</p></li>
<li><p>The ‘score’ parameter filters genes based on their association score with the variants.</p></li>
<li><p>‘output’ can be set to ‘genes’ to return a list of genes or ‘all’ to return detailed information in a DataFrame.</p></li>
</ul>
</section>
<section id="mapping">
<h1>Mapping<a class="headerlink" href="#mapping" title="Permalink to this heading">¶</a></h1>
<p><strong>An utility of genopyc library is the capability of mapping gene Ids between different vocabularies such as ensembl, entrez, Uniprot and gene symbol. It can be easily done with:</strong></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">gene_mapping_many</span><span class="p">(</span><span class="n">query_list</span><span class="p">,</span><span class="n">source</span><span class="p">,</span><span class="n">target</span><span class="p">)</span>
</pre></div>
</div>
<p>Maps gene identifiers from one Id to another.</p>
<p>Args:
query_list (list): List of gene identifiers to be mapped.
source (str): Source Id of the gene identifiers.
target (str): Target Id to which the gene identifiers will be mapped.</p>
<p>Returns:
list: List of gene identifiers mapped to the target namespace.</p>
<p>Raises:
AssertionError: If input query_list contains non-integer values when source or target is ‘entrez’.</p>
<p><strong>To do this genopyc relies on ncbi id mapping and and Uniprot gene info files. The datasets can be updated to the most recent version by running:</strong></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">update_mapping_datasets</span><span class="p">()</span>
</pre></div>
</div>
<p><strong>Moreover, genopyc handles variant Id mapping through the function:</strong></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">convert_variants</span><span class="p">(</span><span class="n">list_of_variants</span><span class="p">,</span> <span class="n">source</span><span class="o">=</span><span class="s1">&#39;variantid&#39;</span><span class="p">,</span> <span class="n">target</span><span class="o">=</span><span class="s1">&#39;rsid&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>Convert variants from one identifier type to another using the Open Targets Genetics API.</p>
<p>Parameters:</p>
<ul class="simple">
<li><p>list_of_variants (list): List of variant identifiers to be converted.</p></li>
<li><p>source (str, optional): Identifier type of the variants in list_of_variants. Default is ‘variantid’.</p></li>
<li><p>target (str, optional): Identifier type to which the variants will be converted. Default is ‘rsid’.</p></li>
</ul>
<p>Returns:</p>
<ul class="simple">
<li><p>list: List of converted variant identifiers.</p></li>
</ul>
<p>Note:</p>
<ul class="simple">
<li><p>Currently supports conversion between ‘variantid’ and ‘rsid’.</p></li>
<li><p>If a variant cannot be converted, it will be skipped, and a message will be printed.</p></li>
</ul>
</section>
<section id="functional-enrichment">
<h1>Functional Enrichment<a class="headerlink" href="#functional-enrichment" title="Permalink to this heading">¶</a></h1>
<p><strong>Once variants have been connected to genes, a functional enrichment can be performed in order to understand which pathways could be perturbated from the variants. We exploit GProfiler <sup>4</sup> to carry out the enrichment analysis through the function:</strong></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">plot_enrichment_analysis_network</span>
</pre></div>
</div>
<p>Perform enrichment analysis and visualize the results as a network where nodes are the functions, size of the nodes is proportional to the number of genes involved in the specific function,
and edges exist between nodes if the functions share genes. The thickness of the edge is proportional to the number of genes involved in the specific function.</p>
<p>Parameters:</p>
<ul class="simple">
<li><p>list_of_genes (list): List of genes for enrichment analysis and network visualization.</p></li>
<li><p>pvalue (float): P-value threshold for significance.</p></li>
<li><p>colormap (str): Colormap for visualization (default: ‘cividis’).</p></li>
<li><p>edgecolor (str): Color of the edges (default: ‘red’).</p></li>
<li><p>mkcolor (str): Color of the markers (default: ‘grey’).</p></li>
<li><p>mkfsize (int): Marker size (default: 10000).</p></li>
<li><p>layout (str): Layout algorithm for network visualization (default: ‘spring’).</p></li>
<li><p>mklinewidths (int): Line width of markers (default: 2).</p></li>
<li><p>alpha (float): Transparency of markers (default: 1).</p></li>
<li><p>figsize (tuple): Figure size (default: (40,20)).</p></li>
<li><p>savefig (bool): Whether to save the figure (default: False).</p></li>
<li><p>factor (int): Scaling factor for the layout (default: 1).</p></li>
<li><p>k (int): Optimal distance between nodes for spring layout (default: 10).</p></li>
<li><p>cbarfontsize (int): Font size of color bar (default: 10).</p></li>
<li><p>labelling (bool): Whether to label the nodes (default: True).</p></li>
<li><p>legend (bool): Whether to show legend (default: False).</p></li>
<li><p>legend_fontsize (int): Font size of legend (default: 20).</p></li>
<li><p>legend_titlefontsize (int): Font size of legend title (default: 25).</p></li>
<li><p>legend_location (tuple): Location of legend (default: (0.5,0.0)).</p></li>
<li><p>legend_col (int): Number of columns in legend (default: 6).</p></li>
<li><p>legend_labelspacing (float): Spacing between legend labels (default: 1.5).</p></li>
<li><p>legend_title (str): Title of legend (default: ‘’).</p></li>
<li><p>legend_columnspacing (float): Spacing between legend columns (default: 1.5).</p></li>
<li><p>legend_handlelength (int): Length of legend handles (default: 3).</p></li>
<li><p>size_legend_nofelements (int): Number of elements in size legend (default: 3).</p></li>
<li><p>cbar_orientation (str): Orientation of color bar (default: ‘horizontal’).</p></li>
<li><p>cbar_loc (tuple): Location of color bar (default: (1, 0.5)).</p></li>
<li><p>method_of_correction (str): Method of multiple testing correction (default: ‘bonferroni’).</p></li>
<li><p>no_evidences (bool): Exclude electronic annotations (default: False).</p></li>
<li><p>no_iea (bool): Exclude Inferred from Electronic Annotation (IEA) evidence (default: True).</p></li>
<li><p>**kwargs: Additional keyword arguments.</p></li>
</ul>
<p>Returns:</p>
<ul class="simple">
<li><p>Pandas DataFrame of the enriched functions</p></li>
</ul>
<p><strong>The returned plot is a network where the nodes are the function.</strong>
<br />
<strong>A new functionality implemented in genopyc is the interactive visualization of the function enrichment. In this network the nodes are proteins and an edge exists between the proteins if there is an interaction. A dropdown menu listing all the enriched functions is utlized in order to highlight the proteins associated to that specific pathway.</strong><br />
<strong>A video showing an example is available in the <a class="reference external" href="https://github.com/freh-g/genopyc/blob/main/function_enrichment_visualization_tool_tutorial_supplementary.mp4">GitHub repository</a></strong></p>
<p><strong>REFERENCES</strong></p>
<p>[1] Piñero, J., Ramírez-Anguita, J. M., Saüch-Pitarch, J., Ronzano, F., Centeno, E., Sanz, F., &amp; Furlong, L. I. (2019). The DisGeNET knowledge platform for disease genomics: 2019 update. Nucleic Acids Research, 48(D1), D845–D855. https://doi.org/10.1093/nar/gkz1021</p>
<p>[2] McLaren, W., Gil, L., Hunt, S. E., Riat, H. S., Ritchie, G. R. S., Thormann, A., Flicek, P., &amp; Cunningham, F. (2016). The Ensembl Variant Effect Predictor. Genome Biology, 17(1). https://doi.org/10.1186/s13059-016-0974-4</p>
<p>[3] Mountjoy, E., Schmidt, E. M., Carmona, M., Schwartzentruber, J., Peat, G., Miranda, A., Fumis, L., Hayhurst, J., Buniello, A., Karim, M. A., Wright, D., Hercules, A., Papa, E., Fauman, E. B., Barrett, J. C., Todd, J. A., Ochoa, D., Dunham, I., &amp; Ghoussaini, M. (2021). An open approach to systematically prioritize causal variants and genes at all published human GWAS trait-associated loci. Nature Genetics, 53(11), 1527–1533. https://doi.org/10.1038/s41588-021-00945-5</p>
<p>[4] Raudvere, U., Kolberg, L., Kuzmin, I., Arak, T., Adler, P., Peterson, H., &amp; Vilo, J. (2019). g:Profiler: a web server for functional enrichment analysis and conversions of gene lists (2019 update). Nucleic Acids Research, 47(W1), W191–W198. https://doi.org/10.1093/nar/gkz369</p>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">genopyc</a></h1>








<h3>Navigation</h3>
<p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Get variants associated with a trait</a></li>
<li class="toctree-l1"><a class="reference internal" href="#investigate-genomic-locations">Investigate genomic locations</a></li>
<li class="toctree-l1"><a class="reference internal" href="#linkage-disequilibrium">Linkage Disequilibrium</a></li>
<li class="toctree-l1"><a class="reference internal" href="#eqtls">eQTLS</a></li>
<li class="toctree-l1"><a class="reference internal" href="#from-variants-to-genes">From Variants to Genes</a></li>
<li class="toctree-l1"><a class="reference internal" href="#mapping">Mapping</a></li>
<li class="toctree-l1"><a class="reference internal" href="#functional-enrichment">Functional Enrichment</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
      <li>Previous: <a href="installation.html" title="previous chapter">Installation</a></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>document.getElementById('searchbox').style.display = "block"</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2024, Francesco Gualdi.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 6.2.1</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.13</a>
      
      |
      <a href="_sources/usage.md.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>